feat: Release enclave binary

felixfaisal opened the pull request:
I think the **********PR********** might be better served if I can provide a full context and the documentation will generally be helpful for future iterations: 

- It updates the `yml` file for `create-release-draft.yml` and includes building enclave and packaging it as an artifact
- It updates the `generate-release-notes.sh` to include the `release notes` for the enclave such as the `rustc_version` , `mrenclave` etc.

So currently, I'm including only the following information for the `Enclave` Release Notes: 
```
rustc: $RUSTC_VERSION
mrenclave: $MRENCLAVE
upstream_commit: $UPSTREAM_COMMIT
```
Let me know if you'd like me to include anything else in the workflow 

Resolves #1628

felixfaisal committed:
fix: create enclave release testing yml file

felixfaisal committed:
feat: add enclave runner

felixfaisal committed:
fix: update the scripts and yml

felixfaisal committed:
fix: update the release workflow to include sgx source

felixfaisal committed:
fix: trying out a hack

felixfaisal committed:
fix: another hack

felixfaisal committed:
fix: update the artifact name

felixfaisal committed:
fix: add changelong from integritee

felixfaisal committed:
fix: add tooling script for changelog

felixfaisal committed:
fix: change the tooling script

felixfaisal committed:
fix: not using any ref for changelog

felixfaisal committed:
fix: pipeline for enclave release/deploy

felixfaisal committed:
Merge branch 'dev' into 1628-fx-release-enclave

# Conflicts:
#	tee-worker/scripts/litentry/release/generate_release_pkg.sh

felixfaisal committed:
fix: include building package and set release type

felixfaisal committed:
fix: add release notes for enclave inside the release notes script

felixfaisal committed:
fix: add vars for release notes

felixfaisal committed:
fix: change the id

felixfaisal committed:
fix: testing setting variable

felixfaisal committed:
fix: syntax error in yml

felixfaisal committed:
fix: multiple changes for echo

felixfaisal committed:
fix: sharing of variables

felixfaisal committed:
fix: add some more needs and output

felixfaisal committed:
fix: store mrenclave version

felixfaisal committed:
fix: cherry pick mrenclave from output

felixfaisal committed:
fix: apply some hack

felixfaisal committed:
fix: apply some hackv2

felixfaisal committed:
fix: apply make mrenclave

felixfaisal committed:
fix: update release notes

felixfaisal committed:
fix: update release yml

felixfaisal committed:
fix: remove test GHA

BillyWooo commented on scripts/generate-release-notes.sh:
```diff
@@ -143,6 +143,23 @@ sha1sum of genesis wasm   : $GENESIS_WASM_HASH
 fi
 
+# release notes for enclave binary 
+if [ "$2" != "runtime" ] && [ "$2" != "client" ] && [ "$2" != "both"]; then 
```
Why can't have an option of `"$2" == "enclave"` ?

felixfaisal replied:
I kind of tried following the existing pattern in the script, let me elaborate: 
```bash
if [ "$2" != "client" ]; then
  echo "## Runtime" >> "$1"
  ```
  A snippet, wherein `$2` != `client` we build the `Runtime`, So i used the same pattern, wherein if not `client`, `runtime` and `both`, we build the `enclave`. This is not very intuitive imo but would require good refactoring if we were to update existing pattern. 
  CC: @Kailai-Wang

BillyWooo replied:
What I mean is: why not replace the whole line with `if [ "$2" == "enclave" ]; then` ? Otherwise what if another new parameter coming in, what will happen here?

felixfaisal replied:
```
if [ "$2" = "enclave" ] || [ "$2" = "all" ]; then 
``` 
Updated to use this

BillyWooo commented on scripts/generate-release-notes.sh:
```diff
@@ -143,6 +143,23 @@ sha1sum of genesis wasm   : $GENESIS_WASM_HASH
+# release notes for enclave binary 
+if [ "$2" != "runtime" ] && [ "$2" != "client" ] && [ "$2" != "both"]; then 
+   echo "Generating Release Notes for Enclave"
+   local UPSTREAM_COMMIT=$(cat tee-worker/upstream_commit)
```
why need `UPSTREAM_COMMIT` here? Or in another words: Parachain repo commit hash is less important?

felixfaisal replied:
I wasn't sure, what all needs to be included in the `release notes` but I thought `UPSTREAM_COMMIT` could be good to track the progress of our worker against the upstream when we release a version from our side.

BillyWooo replied:
I don't think `UPSTREAM_COMMIT` should be included in the release note. It's mainly used for internal usage to track against the upstream worker. I would suggest the current repo git commit hash is more important and valuable. Take the short hash value please.

felixfaisal replied:
Removed ✅

BillyWooo reviewed (commented):
The flow looks fine. Have you got a chance to test on `prod-tee-1` ?

Kailai-Wang commented on .github/workflows/create-release-draft.yml:
```diff
@@ -13,6 +13,11 @@ on:
         description: runtime
         required: true
         default: true
+      release_enclave:
```
Does it always contain `enclave.signed.so` and the worker binary?

BillyWooo replied:
Don't get what you mean here. The package will be created by `./tee-worker/scripts/litentry/release/generate_release_pkg.sh`. So `enclave.signed.so` and worker binary are created together.

Kailai-Wang replied:
Is there any case we only want to release one of them?

Similar to substrate client and runtime

felixfaisal replied:
Yes, if you only select `enclave`, it will release only the `enclave` 
```
          elif [ "${{ github.event.inputs.release_enclave }}" = "true" ]; then
            release_type="enclave"
          elif [ "${{ github.event.inputs.release_enclave }}" = "true" ] && [ "${{ github.event.inputs.release_client }}" = "true" ] && [ "${{ github.event.inputs.release_runtime }}" = "true" ]; then
            release_type="all"
```

Kailai-Wang replied:
It will only release `enclave` but the enclave still contains both binary (integritee-worker) and enclave (.so), right? I mean is there any case we only want to release one of them. Like only the worker part but the enclave doesn't change.

That's why I said "similar to substrate client and runtime".

Not saying we should differentiate them, I'm just raising the question

felixfaisal replied:
I see, As of now, it doesn't do that. Both `integritee-worker` and `enclave` are packaged together.

BillyWooo replied:
@felixfaisal I don't think you answered @Kailai-Wang 's question. He wanted to know: would we have situation that only 'worker' or 'enclave.signed.so' (one of them) needed to be released?
TBH, it could happen. But I would say very rare. 
And we can build them separately (it's easy to do), we just don't do it. Normally we build all together.

Kailai-Wang commented on .github/workflows/create-release-draft.yml:
```diff
@@ -148,6 +157,42 @@ jobs:
             ${{ env.GENESIS_RELEASE }}-genesis-wasm
 
+  ## Build the enclave and package config files
+  generate_package_and_artifact:
```
maybe something like `build-enclave` would be more appropriate for the job name

felixfaisal replied:
Done ✅

Kailai-Wang commented on .github/workflows/create-release-draft.yml:
```diff
@@ -277,6 +322,8 @@ jobs:
       - name: Generate release notes
         run: |
+          export RUSTC_VERSION=${{ needs.generate_package_and_artifact.outputs.rustc_version }}
+          export MRENCLAVE_OUTPUT=${{ needs.generate_package_and_artifact.outputs.mrenclave }}
```
I don't see them in the `needs` list?

felixfaisal replied:
Included it now 😅

Kailai-Wang reviewed (commented):
Thank you, looks good for the most part.

The most important question: have you tried it? :) Do you have something like a generated release draft?

felixfaisal committed:
refactor: remove upstream commit

felixfaisal committed:
refactor: if condition logic

felixfaisal committed:
fix: update job name to build-enclave and set it in the needs

Kailai-Wang commented:
> Do you have something like a generated release draft?

ping 🔼

felixfaisal commented:
So testing this GHA has been difficult to do independently, because it requires building and releasing, and mocking it was tough, However I've tested some parts independently, 
1. Building the enclave and set it artifacts and outputs 
- https://github.com/litentry/litentry-parachain/actions/runs/5551356714/jobs/10137391942 
2. Generating Release Scripts: 
- https://github.com/felixfaisal/litentry-parachain/actions/runs/5543051626/jobs/10118562372 (Not a good mock-up) 

I could not find a better way to test these changes, Let me know if I can test it elsewhere 
CC: @Kailai-Wang @BillyWooo

Kailai-Wang commented:
You simply run https://github.com/litentry/litentry-parachain/actions/workflows/create-release-draft.yml on your feature branch

felixfaisal committed:
fix: update tag for build-wasm

Kailai-Wang commented:
btw it's a bit strange that building enclave only took 6min: https://github.com/litentry/litentry-parachain/actions/runs/5610322811

Maybe it reuses the existing code.
We'd better checkout the repo to a different folder (like `mktemp -d `) and do the compilation ...

felixfaisal committed:
tmp: remove github step for release issue

felixfaisal committed:
tmp: remove ts-tests for testing

felixfaisal committed:
fix: exporting the mrenclave and needs for build-docker

felixfaisal committed:
fix: flow of if conditions

felixfaisal committed:
fix: refactor local var in bash

felixfaisal committed:
fix: checkbox for enclave

felixfaisal committed:
fix: add tar file of enclave binary in the release notes

felixfaisal committed:
refactor: remove unexpected EOF

felixfaisal committed:
fix: add fi for missed if/condition

felixfaisal committed:
fix: add artifact of enclave in release notes

felixfaisal committed:
fix: artifact path

felixfaisal committed:
fix: tee-worker artifact upload pattern

felixfaisal committed:
fix: test building only enclave

felixfaisal committed:
fix: downloading specific tee-worker zip

felixfaisal committed:
fix: download enclave separate

felixfaisal committed:
fix: debug files in GHA

felixfaisal committed:
fix: add tee-worker release

felixfaisal committed:
fix: conditons for create-release job

felixfaisal committed:
fix: remove hardcoded values

felixfaisal committed:
refactor: remove typo

felixfaisal committed:
fix: if/else flow for independent enclave build

Kailai-Wang commented:
Whenever it's ready for review, please attach a link to the **release draft** for reference, thanks :)

BillyWooo commented on .github/workflows/create-release-draft.yml:
```diff
@@ -148,43 +158,43 @@ jobs:
             ${{ env.GENESIS_RELEASE }}-genesis-state
             ${{ env.GENESIS_RELEASE }}-genesis-wasm
 
-  ## test again the built docker image ##
```
Why we simply remove these testing steps?

felixfaisal replied:
Hey @BillyWooo I removed them because it was failing and not related to the changes I've made, So that I can only test the changes I've made. I will include them back.

felixfaisal committed:
fix: remove downloading specific artifact

felixfaisal commented:
So I had a few succesfull tests completed 
- https://github.com/litentry/litentry-parachain/actions/runs/5660062718/job/15338016011 (both) 
- https://github.com/litentry/litentry-parachain/actions/runs/5655699432/job/15323894884 (all) 
- https://github.com/litentry/litentry-parachain/actions/runs/5654230410 (enclave) 

I will add the steps I removed for testing

felixfaisal committed:
fix: include steps removed for testing

BillyWooo reviewed (approved):
thanks

Kailai-Wang reviewed (changes requested):
A few questions:

- why was the `release-xxx.tar.gz` file still included in the [release](https://github.com/litentry/litentry-parachain/releases/tag/untagged-89b09d8665e2a4b87955) with `client` and `runtime` only?
- the `rustc` version is wrong - you need the one from the `tee-worker/` folder
- can you align the indention with the existing dependencies? (like polkadot dependency)
- we should include the hash of the `binary` and `enclave.signed.so` too
- did you differentiate between a `dev` build and `prod` build?

felixfaisal committed:
fix: make changes based on comments

felixfaisal committed:
fix: update FILENAME

felixfaisal commented:
> A few questions:
> 
> * why was the `release-xxx.tar.gz` file still included in the [release](https://github.com/litentry/litentry-parachain/releases/tag/untagged-89b09d8665e2a4b87955) with `client` and `runtime` only?
> * the `rustc` version is wrong - you need the one from the `tee-worker/` folder
> * can you align the indention with the existing dependencies? (like polkadot dependency)
> * we should include the hash of the `binary` and `enclave.signed.so` too
> * did you differentiate between a `dev` build and `prod` build?

 why was the release-xxx.tar.gz file still included in the [release](https://github.com/litentry/litentry-parachain/releases/tag/untagged-89b09d8665e2a4b87955) with client and runtime only?
It was included because of this step that I had left behind, which I removed in this commit - https://github.com/litentry/litentry-parachain/pull/1892/commits/4f1a539b6f7bd3a859bf5d66d8366b98b952af90

I've incorporated all the other changes, except for 
> * did you differentiate between a `dev` build and `prod` build?

For now the binaries are built on production mode, because `generate_relase_pkg.sh` builds it on Production, Should I create an input to allow us to chose whether production or dev? 

This is the recent run: 
https://github.com/litentry/litentry-parachain/releases/tag/untagged-f0f321cd562d7eebdbfb

Kailai-Wang commented:
Thanks!

Can you rerun it with the "previous release" set? I want to check the changelog and collected PRs.

felixfaisal commented:
Here's the draft with the previous release having changelog  - https://github.com/litentry/litentry-parachain/releases/tag/untagged-8229ac39f97c3c61c335

Kailai-Wang commented:
Looks good!
Better would be that we list those PRs with C0 and C1 first to highlight them even more, which are followed by the PRs without any tags, ideally with toggle list.

But that should be out of this PR's scope.

Kailai-Wang reviewed (approved):
I approve it - one minor thing is we need shasum for both binary and enclave. In this PR there's only one, which is not enough, and please mind the upper/lower case letters :) (to be consistent with other metadata)

felixfaisal committed:
fix: add sha256 for enclave and binary and align

felixfaisal committed:
debug: files in release tar

felixfaisal committed:
debug: checking files

felixfaisal committed:
debug: check all files inside tmp

felixfaisal committed:
debug: finding enclave and binary

felixfaisal committed:
fix: sha256sum for enclave and worker

felixfaisal committed:
fix: rename binaries

felixfaisal committed:
refactor: remove unused var
